swerve_controller/test/include/swerve_controller/swervebot.h

```diff
@@ -0,0 +1,210 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Exobotic nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+
+#ifndef SWERVE_CONTROLLER_SWERVEBOT_H
+#define SWERVE_CONTROLLER_SWERVEBOT_H
+
+#include <cmath>
+#include <gtest/gtest.h>
+#include <ros/ros.h>
+
+#include <geometry_msgs/Twist.h>
+#include <nav_msgs/Odometry.h>
+#include <tf/tf.h>
+
+#include <std_srvs/Empty.h>
+
+// Floating-point value comparison threshold
+const double EPS = 0.02;
+const double POSITION_TOLERANCE = 0.1;  // 10cm/5s = 2 cm-s precision
+const double ORIENTATION_TOLERANCE = 0.2;  // 0.2rad = 11 degrees precision
+const double VELOCITY_TOLERANCE = 0.05;  // not consistent (linear and angular speed)
+
+
+
+/**
+ * This class allows to test the swerve controller with google test
+ */
+class SwerveControllerTest : public ::testing::Test
+{
+    public:
+        /**
+         * \brief Constructor. Create services and topic publishers and subscribers
+         */
+        SwerveControllerTest()
+        : received_first_odom(false)
+        , cmd_twist_pub(nh.advertise<geometry_msgs::Twist>("cmd_vel", 100))
+        , odom_sub(nh.subscribe("odom", 100, &SwerveControllerTest::odomCallback, this))
+        , start_srv(nh.serviceClient<std_srvs::Empty>("start"))
+        , stop_srv(nh.serviceClient<std_srvs::Empty>("stop"))
+        , loop_rate(ros::Rate(40))
+        {
+        }
+
+        /**
+         * \brief Desctructor. Shuts off the subscriber
```

Comment:
subscribers know how to shut down on their own usually

Reply:
I've just copied your [code](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/four_wheel_steering_controller/test/src/test_common.h) and commented it for my own understanding. I will remove it.